R/auth.R

```diff
@@ -0,0 +1,292 @@
+  if (is.null(.state$auth)) {
+    auth_as()
+  }
+  .state$auth
+}
+
+# Save authentication across sessions -------------------------------------
+
+#' Save an authentication mechanism for use in a future session
+#' 
+#' Use `auth_save()` with [auth_as()] to avoid repeatedly entering app 
+#' credentials, making it easier to share auth between projects.
+#' 
+#' @param auth One of [rtweet_app()], [rtweet_bot()], or [rtweet_user()].
+#' @param name Cache name to use.
+#' @export
+#' @examples 
+#' \dontrun{
+#' # save app auth for use in other sessions
+#' auth <- rtweet_app()
+#' auth_save(auth, "my-app")
+#' 
+#' # later, in a different session...
+#' auth_us("my-app")
+#' }
+auth_save <- function(auth, name) {
+  stopifnot(is_auth(auth), is_string(name))
+  
+  path <- auth_path(paste0(name, ".rds"))
+  inform(paste0("Saving auth to '", path, "'"))
+  
+  dir.create(auth_path(), showWarnings = FALSE, recursive = TRUE)
+  saveRDS(auth, path)
+  invisible(path)
+}
+
+auth_path <- function(...) {
+  # Use private option to make testing easier
+  path <- getOption("rtweet:::config_dir", rappdirs::user_config_dir("rtweet", "R"))
+  file.path(path, ...)
+}
+
+# Set default auth -------------------------------------------------------------
+
+#' Set default authentication
+#' 
+#' `auth_as()` sets up the default authentication mechanism used by all 
+#' rtweet API calls. See [rtweet_user()] to learn more about the three
+#' available authentication options.
+#' 
+#' @param auth One of the following options:
+#'   * `NULL`, the default, will look for cached authentication from a previous
+#'      session. If none are found, will automatically set up [rtweet_user()];
+#'      if one is found, will use that; if multiple are found, it will error.
+#'   * A string giving the name of a cached auth file.
+#'   * An object created by [rtweet_app()], [rtweet_bot()], or [rtweet_user()].
+#' @return Invisibly returns the previous authentication mechanism.
+#' @examples 
+#' \dontrun{
+#' # Use app auth for the remainder of this session:
+#' my_app <- rtweet_app()
+#' auth_as(my_app)
+#' 
+#' # Switch back to the default user based auth
+#' auth_as()
+#' 
+#' # Load auth saved by auth_save()
+#' auth_as("my-saved-app")
+#' }
+auth_as <- function(auth = NULL) {
+  old <- .state$auth
+  .state$auth <- find_auth(auth)
+  invisible(old)
+}
+
+find_auth <- function(auth = NULL) {
+  if (is.null(auth)) {
+    if (is_testing()) {
+      rtweet_test() %||% no_token()
+    } else if (is_dev_mode()) {
+      rtweet_test() %||% default_cached_auth()
+    } else{
+      default_cached_auth()
+    }
+  } else if (is_auth(auth)) {
+    auth
+  } else if (is_string(auth)) {
+    path <- auth_path(paste0(auth, ".rds"))
+    if (!file.exists(path)) {
+      abort(paste0("Can't find saved auth with name '", auth, "'"))
+    }
+    inform(paste0("Reading auth from '", path, "'"))
+    readRDS(path)
+  } else {
+    abort("Unrecognised input to `auth`")
+  }
+}
```

Comment:
I find it hard to follow, why there is a specific branch for dev_mode?
On my .Rprofile I set to load devtools, so it always detect that I'm on dev_mode and returns the rtweet_test(). Not sure how widespread this is but I expect all potential contributors and developers of dependencies to have loaded devtools. Perhaps it should test if it is on dev_mode by looking at the dev_path option set up by `devtools::dev_mode()`.

Reply:
This only applies if you have loaded **rtweet** with devtools, and thus prevents you running the tests with your personal account.